extensions/PhabBugz/lib/Util.pm

```diff
@@ -305,4 +306,135 @@ sub set_phab_user {
+    }
+  }
+
+  if (!$blocking_project) {
+    INFO("$rev_identifier: Blocking reviewer project not found.");
+    return;
+  }
+
+  INFO( "$rev_identifier: Blocking reviewer project "
+      . $blocking_project->name
+      . ' found.');
+
+  # 2. Once the blocking reviewer group is determined, query Phabricator for
+  # list of group members and match up the BMO user account. Sort them by user
+  # id descending.
+  my $project_members
+    = [sort { $a->id <=> $b->id } @{$blocking_project->members}];
+
+  # 3. Make sure that none of the individual group members are not already
+  # set as a blocking reviewer. If so, then remove the blocking group and return.
+  foreach my $member (@{$project_members}) {
+    if (any { $_->id == $member->id } @blocking_users) {
+      INFO(
+        "$rev_identifier: Member of blocking reviewer project already set as a reviewer. Removing blocking reviewer project."
+      );
+      $revision->remove_reviewer($blocking_project->phid);
+      $revision->update;
+      return;
+    }
+  }
+
+  # 4. Going in order, look up in the phab_reviewer_rotation table for each
+  # user to see if they are already a reviewer on another attachment.
+  my $dbh = Bugzilla->dbh;
+
+  my $found_reviewer;
+  foreach my $member (@{$project_members}) {
+
+   # 5. If the user has a revison they are reviewing currently, load the revision
+   # details and check if it is closed. If it is, then clear the row from the table.
+    my $rev_phid = $dbh->selectrow_array(
+      'SELECT revision_phid FROM phab_reviewer_rotation WHERE project_phid = ? AND user_phid = ?',
+      undef, $blocking_project->phid, $member->phid
+    );
+
+    if ($rev_phid) {
+      INFO("$rev_identifier: Previous review found for " . $member->id);
+
+      my $rev_obj = Bugzilla::Extension::PhabBugz::Revision->new_from_query(
+        {phids => [$rev_phid]});
+
+# 6. If the user is already a reviewer or they were but their revision is now closed, skip to the next
+# user in the list.
+      if ($rev_obj->status eq 'closed' || $rev_obj->status eq 'abandoned') {
+        INFO(
+          "$rev_identifier: Previous reviewers revision is closed or abandoned. Removing reviewer from database and skipping."
+        );
+        $dbh->do(
+          'DELETE FROM phab_reviewer_rotation WHERE revision_phid = ? AND project_phid = ? AND user_phid = ?',
+          undef, $rev_phid, $blocking_project->phid, $member->phid
+        );
+      }
+
+      next;
+    }
+
+ # 7. Once a potential reviewer has been found, look to see if they can see the bug,
+ # and they are not set to away (not accepting reviews). If both are are negative,
+ # we choose the next person in the list.
+    if ( $member->bugzilla_user->can_see_bug($revision->bug->id)
+      && $member->bugzilla_user->settings->{block_reviews}->{value} ne "on")
+    {
+      INFO("$rev_identifier: Found new reviewer " . $member->id);
+      $found_reviewer = $member;
+      last;
+    }
+  }
+
+  if ($found_reviewer) {
```

Comment:
Looking at this logic, it seems it is possible for this system to *not* find a reviewer, and thus leave the blocking review group set on the revision. AIUI the goal of this system is to always set a specific person as the reviewer on the revision when a blocking review group is set. We should have some logic here to assign a reviewer even if one isn't found with the current logic.